File: lobster-input-ios/LobsterKeyboard/Typing/Accessibility/TypingAccessibilityLabels.swift
import UIKit

/// 键盘无障碍标签与状态播报(集中管理,避免散落)。
enum TypingAccessibilityLabels {

    static func keyLabel(
        for spec: TypingKeySpec,
        shiftActive: Bool,
        capsLock: Bool
    ) -> String {
        switch spec.type {
        case .letter:
            let ch = (!shiftActive && !capsLock) ? spec.main : spec.main.uppercased()
            return MobileStrings.a11yKeyLetter(ch)
        case .t9:
            return MobileStrings.a11yKeyT9(main: spec.main, sub: spec.sub ?? "")
        case .delete:
            return MobileStrings.a11yKeyDelete()
        case .space:
            return MobileStrings.a11yKeySpace()
        case .enter:
            return MobileStrings.a11yKeyEnter()
        case .shift:
            if capsLock { return MobileStrings.a11yKeyCapsLock() }
            if shiftActive { return MobileStrings.a11yKeyShiftOn() }
            return MobileStrings.a11yKeyShift()
        case .lang:
            return MobileStrings.a11yKeyLang()
        case .layout:
            return MobileStrings.a11yKeyLayout()
        case .symbol, .num:
            return MobileStrings.a11yKeySymbols()
        case .alpha:
            return MobileStrings.a11yKeyAlpha()
        case .symPage:
            return MobileStrings.a11yKeySymPage()
        case .symChar:
            return MobileStrings.a11yKeyChar(spec.main)
        // 模式轮换键:读出下一模式(spec.sub);符号板入口读符号页通用描述
        case .modeCycle:
            return MobileStrings.a11yKeyLayout() + (spec.sub.map { " \($0)" } ?? "")
        case .symBoard:
            return MobileStrings.a11yKeySymbols()
        case .syllable:
            return MobileStrings.syllableSepKey()
        case .undo:
            return MobileStrings.undo()
        case .mic:
            return MobileStrings.a11yKeyMic()
        // 9 宫格 1 键(@#):高频符号候选入口,读作符号键
        case .symCands:
            return MobileStrings.a11yKeySymbols()
        case .gap:
            return ""
        }
    }

    static func candidateLabel(_ candidate: PinyinCandidate, index: Int) -> String {
        "\(index + 1). \(candidate.word)"
    }

    static func announceStatus(on view: UIView, message: String) {
        view.accessibilityLabel = message
        UIAccessibility.post(notification: .announcement, argument: message)
    }
}

File: lobster-input-android/app/src/main/java/com/lobster/input/keyboard/typing/accessibility/TypingAccessibilityLabels.kt
package com.lobster.input.keyboard.typing.accessibility

import android.content.Context
import android.view.View
import android.view.accessibility.AccessibilityEvent
import com.lobster.input.core.locale.MobileStrings
import com.lobster.input.keyboard.pinyin.Candidate
import com.lobster.input.keyboard.typing.KeySpec
import com.lobster.input.keyboard.typing.KeyType

/** 键盘无障碍标签与状态播报(集中管理,避免散落)。 */
object TypingAccessibilityLabels {

    fun applyKeyLabel(view: View, spec: KeySpec, context: Context, shiftActive: Boolean, capsLock: Boolean) {
        view.contentDescription = when (spec.type) {
            KeyType.LETTER -> {
                val ch = if (shiftActive || capsLock) spec.main.uppercase() else spec.main
                MobileStrings.a11yKeyLetter(context, ch)
            }
            KeyType.T9 -> MobileStrings.a11yKeyT9(context, spec.main, spec.sub ?: "")
            KeyType.DELETE -> MobileStrings.a11yKeyDelete(context)
            KeyType.SPACE -> MobileStrings.a11yKeySpace(context)
            KeyType.ENTER -> MobileStrings.a11yKeyEnter(context)
            KeyType.SHIFT -> if (capsLock) MobileStrings.a11yKeyCapsLock(context)
            else if (shiftActive) MobileStrings.a11yKeyShiftOn(context)
            else MobileStrings.a11yKeyShift(context)
            KeyType.LANG -> MobileStrings.a11yKeyLang(context)
            KeyType.LAYOUT -> MobileStrings.a11yKeyLayout(context)
            KeyType.SYMBOL, KeyType.NUM -> MobileStrings.a11yKeySymbols(context)
            KeyType.ALPHA -> MobileStrings.a11yKeyAlpha(context)
            KeyType.SYM_PAGE -> MobileStrings.a11yKeySymPage(context)
            KeyType.SYM_CHAR -> MobileStrings.a11yKeyChar(context, spec.main)
            // 模式轮换键:读出下一模式(spec.sub);符号板入口读布局键通用描述
            KeyType.MODE_CYCLE -> MobileStrings.a11yKeyLayout(context) + (spec.sub?.let { " $it" } ?: "")
            KeyType.SYM_BOARD -> MobileStrings.a11yKeySymbols(context)
            // 9 宫格 1 键(@#):高频符号候选入口,读作符号键
            KeyType.SYM_CANDS -> MobileStrings.a11yKeySymbols(context)
            else -> spec.main
        }
    }

    fun announceCandidate(view: View, candidate: Candidate, index: Int) {
        view.contentDescription = "${index + 1}. ${candidate.word}"
    }

    fun announceStatus(view: View, message: String) {
        view.announceForAccessibility(message)
        view.sendAccessibilityEvent(AccessibilityEvent.TYPE_ANNOUNCEMENT)
    }
}
